### PR TITLE
feat(rel): plug upgradetest into release tests

### PR DESCRIPTION
**Pull request by jhchabran**
This PR introduces the `target-registry` and `from-registry` flags. These flags allow the specification of the image registry to pull from when running the upgrade test. `target-registry` sets the `migrator` image image registry as well as the `frontend` during the upgrade steps of the various upgrade test types. `from-registry` sets the image registry used to pull images on test env init -- initializing `pgsql`, `codeintel-db`, `codeinsights-db`, migrator`, and `frontend`. 

## Test plan
The following command works on main --
`bazel run //testing/tools/upgradetest:sh_upgradetest -- all -pv 5.3.9104 -svs 5.2.2 -mvs 5.1.0 -avs 5.1.0`

The command should work with new flag `bazel run //testing/tools/upgradetest:sh_upgradetest -- all -pv 5.3.9104 -svs 5.2.2 -mvs 5.1.0 -avs 5.1.0 -target-registry us-central1-docker.pkg.dev/sourcegraph-ci/rfc795-internal/`

<details>

```
➜  sourcegraph git:(main) bazel run //testing/tools/upgradetest:sh_upgradetest -- all -pv 5.3.9104 -svs 5.2.2 -mvs 5.1.0 -avs 5.1.0
INFO: Analyzed target //testing/tools/upgradetest:sh_upgradetest (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //testing/tools/upgradetest:sh_upgradetest up-to-date:
  bazel-bin/testing/tools/upgradetest/sh_upgradetest
Aspect @@rules_rust//rust/private:clippy.bzl%rust_clippy_aspect of //testing/tools/upgradetest:sh_upgradetest up-to-date (nothing to build)
INFO: Elapsed time: 2.065s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: external/bazel_tools/tools/test/test-setup.sh testing/tools/upgradetest/sh_upgradetest testing/tools/upgradetest/go_upgradetest_/go_upgradetest cmd/migrator/image_tarball.sh cmd/frontend/image_tarball.sh internal/database/_codeinsights_squashed.sql internal/database/_codeintel_squashed.sql internal/database/_frontend_squashed.sql internal/database/_schema.codeinsights.json internal/database/_schema.codeinsights.md internal/database/_schema.codeintel.json internal/database/_schema.codeintel.md internal/database/_schema.json internal/database/_schema.md all -pv 5.3.9104 -svs 5.2.2 -mvs 5.1.0 -avs 5.1.0
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //testing/tools/upgradetest:sh_upgradetest
-----------------------------------------------------------------------------
Loaded image: migrator:candidate
Loaded image: frontend:candidate
Latest stable release version:  5.3.9104
Latest minor version:  5.3.0
Target version:  5.3.9104
Migrator image used to upgrade:  sourcegraph/migrator:5.3.9104
Standard Versions: [5.2.2]
Multiversion Versions: [5.1.0]
Autoupgrade Versions: [5.1.0]
auto:  5.1.0
std:  5.2.2
mvu:  5.1.0
--- 🕵️  Standard Upgrade Tests:
✅ 5.2.2 Passed -- 44.370164625s
--- 🕵️  Multiversion Upgrade Tests:
✅ 5.1.0 Passed -- 47.75924125s
--- 🕵️  Auto Upgrade Tests:
✅ 5.1.0 Passed -- 45.660516292s
```

</details>

### CODE REVIEW

File: release.yaml [hunk 1]
```diff
@@ -62,54 +62,57 @@ internal:
...
-
-        echo "checking migrator image for older schema description files"
-        count=$(docker run -t --rm --entrypoint /bin/sh "${IMAGE}" -c "ls -al /schema-descriptions/*-internal_database_schema*json | wc -l" | tr -d '[:space:]')
-
-        # in our newer releases we have upwards of 300 schema descriptions, so we check that we have at least 300
-        if [[ "$count" -lt 300 ]]; then
-          echo "migrator:{{tag}} does not contain the correct amount of schema description files for this release - expected more than 300 got ${count}"
-          exit 1
-        fi
+        bazel run //testing/tools/upgradetest:sh_upgrade_test -- std
```
[the remark this answers is not in the document]
Comment (DaedalusG) [reply release.yaml:110]:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //testing/tools/upgradetest:sh_upgradetest
-----------------------------------------------------------------------------
Loaded image: migrator:candidate
Loaded image: frontend:candidate
Latest stable release version:  5.3.3
Latest minor version:  5.3.0
Target version:  5.3.9104
Migrator image used to upgrade:  %!s(<nil>)migrator:5.3.9104
Standard Versions: [5.3.2]
std:  5.3.2
```
^^ The base command on our branch

```
➜  sourcegraph git:(main) bazel run //testing/tools/upgradetest:sh_upgradetest -- std -pv 5.3.2 -svs 5.2.2
INFO: Analyzed target //testing/tools/upgradetest:sh_upgradetest (1987 packages loaded, 16639 targets configured).
INFO: From GoLink dev/sg/sg_/sg [for tool]:
ld: warning: ignoring duplicate libraries: '-lm'
INFO: From Running sg migration ...:
+ test -d ''
+ IS_UNIX_POSTGRES=1
+ set -ex
+ '[' 1 -eq 0 ']'
+ export PGDATASOURCE=postgres://sourcegraph:@:/sg-squasher-frontend
+ PGDATASOURCE=postgres://sourcegraph:@:/sg-squasher-frontend
+ export CODEINTEL_PGDATASOURCE=postgres://sourcegraph:@:/sg-squasher-codeintel
+ CODEINTEL_PGDATASOURCE=postgres://sourcegraph:@:/sg-squasher-codeintel
+ export CODEINSIGHTS_PGDATASOURCE=postgres://sourcegraph:@:/sg-squasher-codeinsights
+ CODEINSIGHTS_PGDATASOURCE=postgres://sourcegraph:@:/sg-squasher-codeinsights
+ trap 'dropdb --if-exists sg-squasher-frontend && echo '\''temp db sg-squasher-frontend dropped'\''' EXIT
+ trap 'dropdb --if-exists sg-squasher-codeintel && echo '\''temp db sg-squasher-codeintel dropped'\''' EXIT
+ trap 'dropdb --if-exists sg-squasher-codeinsights && echo '\''temp db sg-squasher-codeinsights dropped'\''' EXIT
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration squash-all -skip-teardown -db frontend -f bazel-out/darwin_arm64-fastbuild/bin/internal/database/_frontend_squashed.sql
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration squash-all -skip-teardown -db codeintel -f bazel-out/darwin_arm64-fastbuild/bin/internal/database/_codeintel_squashed.sql
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration squash-all -skip-teardown -db codeinsights -f bazel-out/darwin_arm64-fastbuild/bin/internal/database/_codeinsights_squashed.sql
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration describe -db frontend --format=json -force -out bazel-out/darwin_arm64-fastbuild/bin/internal/database/_schema.json
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration describe -db codeintel --format=json -force -out bazel-out/darwin_arm64-fastbuild/bin/internal/database/_schema.codeintel.json
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration describe -db codeinsights --format=json -force -out bazel-out/darwin_arm64-fastbuild/bin/internal/database/_schema.codeinsights.json
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration describe -db frontend --format=psql -force -out bazel-out/darwin_arm64-fastbuild/bin/internal/database/_schema.md
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration describe -db codeintel --format=psql -force -out bazel-out/darwin_arm64-fastbuild/bin/internal/database/_schema.codeintel.md
+ bazel-out/darwin_arm64-opt-exec-ST-13d3ddad9198/bin/dev/sg/sg_/sg migration describe -db codeinsights --format=psql -force -out bazel-out/darwin_arm64-fastbuild/bin/internal/database/_schema.codeinsights.md
+ dropdb --if-exists sg-squasher-codeinsights
+ echo 'temp db sg-squasher-codeinsights dropped'
Creating local Postgres database sg-squasher-frontend......
✅ Created local Postgres database sg-squasher-frontend
Migrating PostgreSQL schema......
Attempting connection to frontend...
✅ Connection to frontend succeeded
✅ Migrated PostgreSQL schema
Dumping current database......
✅ Dumped current database
Creating local Postgres database sg-squasher-codeintel......
✅ Created local Postgres database sg-squasher-codeintel
Migrating PostgreSQL schema......
Attempting connection to codeintel...
✅ Connection to codeintel succeeded
✅ Migrated PostgreSQL schema
Dumping current database......
✅ Dumped current database
Creating local Postgres database sg-squasher-codeinsights......
✅ Created local Postgres database sg-squasher-codeinsights
Migrating PostgreSQL schema......
Attempting connection to codeinsights...
✅ Connection to codeinsights succeeded
✅ Migrated PostgreSQL schema
Dumping current database......
✅ Dumped current database
Attempting connection to frontend...
✅ Connection to frontend succeeded
Describing database frontend......
✅ Description of frontend written to target
Attempting connection to codeintel...
✅ Connection to codeintel succeeded
Describing database codeintel......
✅ Description of codeintel written to target
Attempting connection to codeinsights...
✅ Connection to codeinsights succeeded
Describing database codeinsights......
✅ Description of codeinsights written to target
Attempting connection to frontend...
✅ Connection to frontend succeeded
Describing database frontend......
✅ Description of frontend written to target
Attempting connection to codeintel...
✅ Connection to codeintel succeeded
Describing database codeintel......
✅ Description of codeintel written to target
Attempting connection to codeinsights...
✅ Connection to codeinsights succeeded
Describing database codeinsights......
✅ Description of codeinsights written to target
temp db sg-squasher-codeinsights dropped
INFO: Found 1 target...
Target //testing/tools/upgradetest:sh_upgradetest up-to-date:
  bazel-bin/testing/tools/upgradetest/sh_upgradetest
Aspect @@rules_rust//rust/private:clippy.bzl%rust_clippy_aspect of //testing/tools/upgradetest:sh_upgradetest up-to-date (nothing to build)
INFO: Elapsed time: 78.643s, Critical Path: 31.34s
INFO: 539 processes: 3 internal, 536 darwin-sandbox.
INFO: Build completed successfully, 539 total actions
INFO: Running command line: external/bazel_tools/tools/test/test-setup.sh testing/tools/upgradetest/sh_upgradetest testing/tools/upgradetest/go_upgradetest_/go_upgradetest cmd/migrator/image_tarball.sh cmd/frontend/image_tarball.sh internal/database/_codeinsights_squashed.sql internal/database/_codeintel_squashed.sql internal/database/_frontend_squashed.sql internal/database/_schema.codeinsights.json internal/database/_schema.codeinsights.md internal/database/_schema.codeintel.json internal/database/_schema.codeintel.md internal/database/_schema.json internal/database/_schema.md std -pv 5.3.2 -svs 5.2.2
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //testing/tools/upgradetest:sh_upgradetest
-----------------------------------------------------------------------------
The image migrator:candidate already exists, renaming the old one with ID sha256:3e32adeed660d6b2571a2ee173a2419ff6a9231a9c21f60e94459e95343060ad to empty string
Loaded image: migrator:candidate
The image frontend:candidate already exists, renaming the old one with ID sha256:ede35ffb7c370a2c75f6eee443ad35ea420646f38e012952f7e4c13a632fd2d4 to empty string
Loaded image: frontend:candidate
Latest stable release version:  5.3.3
Latest minor version:  5.3.0
Target version:  5.3.2
Migrator image used to upgrade:  sourcegraph/migrator:5.3.2
Standard Versions: [5.2.2]
std:  5.2.2
--- 🕵️  Standard Upgrade Tests:
✅ 5.2.2 Passed -- 39.395950458s
```
^^ Verification the basic usage is working on main

File: testing/tools/upgradetest/main.go [hunk 2]
```diff
@@ -74,7 +90,12 @@ func main() {
 					}
 
 					// Get init versions to use for initializing upgrade environments for tests
-					latestMinorVersion, latestStableVersion, targetVersion, stdVersions, mvuVersions, autoVersions, err := handleVersions(cCtx, cCtx.StringSlice("standard-versions"), cCtx.StringSlice("mvu-versions"), cCtx.StringSlice("auto-versions"), cCtx.String("post-release-version"))
+					latestMinorVersion, latestStableVersion, targetVersion, stdVersions, mvuVersions, autoVersions, err := handleVersions(cCtx,
+						cCtx.StringSlice("standard-versions"),
+						cCtx.StringSlice("mvu-versions"),
+						cCtx.StringSlice("auto-versions"),
+						cCtx.String("post-release-version"),
+					)
```
Comment (DaedalusG) [inline testing/tools/upgradetest/main.go:98]: much cleaner to look at here

File: testing/tools/upgradetest/tests-util.go [hunk 3]
```diff
@@ -530,6 +531,7 @@ func startFrontend(ctx context.Context, test Test, image, version, networkName s
 	errChan := make(chan error)
 	go func() {
 		if _, err := run.Cmd(ctx, cmdString...).Run().String(); err != nil {
+			fmt.Println("🚨 failed to start frontend: ", err)
```
Comment (DaedalusG) [inline testing/tools/upgradetest/tests-util.go:534]: @jhchabran frontend will always throw a kubernetes error here, I think its baked into the frontend container that it expects some kubernetes config env vars
```
{"SeverityText":"ERROR","Timestamp":1712550496982873301,"InstrumentationScope":"frontend.configurationSource","Caller":"cli/config.go:584","Function":"github.com/sourcegraph/sourcegraph/cmd/frontend/internal/cli.serviceConnections","Body":"failed to get zoekt endpoints for service connections","Resource":{"service.name":"frontend","service.version":"5.3.9104","service.instance.id":"6509353bdf30"},"Attributes":{"error":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}}
--- 🕵️  Standard Upgrade Tests:
✅ 5.2.2 Passed -- 42.734448042s
--- 🕵️  Multiversion Upgrade Tests:
✅ 5.1.0 Passed -- 55.510806416s
--- 🕵️  Auto Upgrade Tests:
✅ 5.1.0 Passed -- 43.542348583s
```
^^ you can see here frontend does start without errors, for debugging the best way to see container errors is like so 
```
					// This is where we do the majority of our printing to stdout.
					results.OrderByVersion()
					results.PrintSimpleResults()
					results.DisplayErrors()
```

**Comment by DaedalusG** [comment]
It looks like the test can't pull from our registry
`➜  sourcegraph git:(wgjh/rel/upgradetest) bazel run //testing/tools/upgradetest:sh_upgradetest -- all -pv v5.3.9104 -svs 5.2.2 -mvs 5.1.0 -avs 5.1.0 -target-registry sourcegraph/ -from-registry sourcegraph/`
```
Loaded image: migrator:candidate
Loaded image: frontend:candidate
👉 Upgrade test ...
Latest stable release version:  5.3.9104
Latest minor version:  5.3.0
Target version:  5.3.9104
Migrator image used to upgrade:  sourcegraph/migrator:5.3.9104
Standard Versions: [5.2.2]
Multiversion Versions: [5.1.0]
Autoupgrade Versions: [5.1.0]
auto:  5.1.0
mvu:  5.1.0
std:  5.2.2
--- 🕵️  Standard Upgrade Tests:
✅ 5.2.2 Passed -- 41.835429667s
--- 🕵️  Multiversion Upgrade Tests:
✅ 5.1.0 Passed -- 45.461936959s
--- 🕵️  Auto Upgrade Tests:
✅ 5.1.0 Passed -- 53.732386917s
```
`➜  sourcegraph git:(wgjh/rel/upgradetest) bazel run //testing/tools/upgradetest:sh_upgradetest -- all -pv 5.3.9104 -svs 5.2.2 -mvs 5.1.0 -avs 5.1.0 -target-registry us-central1-docker.pkg.dev/sourcegraph-ci/rfc795-internal/ -from-registry sourcegraph/`
```
Loaded image: migrator:candidate
Loaded image: frontend:candidate
👉 Upgrade test ...
Latest stable release version:  5.3.9104
Latest minor version:  5.3.0
Target version:  5.3.9104
Migrator image used to upgrade:  us-central1-docker.pkg.dev/sourcegraph-ci/rfc795-internal/migrator:5.3.9104
Standard Versions: [5.2.2]
Multiversion Versions: [5.1.0]
Autoupgrade Versions: [5.1.0]
auto:  5.1.0
mvu:  5.1.0
std:  5.2.2
🚨 failed to stop frontend after testing:  exit status 1: Error response from daemon: No such container: auto_frontend_34794dc2
🚨 failed to remove frontend after testing:  exit status 1: Error response from daemon: No such container: auto_frontend_34794dc2
--- 🕵️  Standard Upgrade Tests:
🚨 5.2.2 Failed -- 25.398653708s
🚨 failed to upgrade: exit status 125: Unable to find image 'us-central1-docker.pkg.dev/sourcegraph-ci/rfc795-internal/migrator:5.3.9104' locally
docker: Error response from daemon: Head "https://us-central1-docker.pkg.dev/v2/sourcegraph-ci/rfc795-internal/migrator/manifests/5.3.9104": denied: Unauthenticated request. Unauthenticated requests do not have permission "artifactregistry.repositories.downloadArtifacts" on resource "projects/sourcegraph-ci/locations/us-central1/repositories/rfc795-internal" (or it may not exist).
See 'docker run --help'.
--- 🕵️  Multiversion Upgrade Tests:
🚨 5.1.0 Failed -- 25.732303083s
🚨 failed to upgrade: exit status 125: Unable to find image 'us-central1-docker.pkg.dev/sourcegraph-ci/rfc795-internal/migrator:5.3.9104' locally
docker: Error response from daemon: Head "https://us-central1-docker.pkg.dev/v2/sourcegraph-ci/rfc795-internal/migrator/manifests/5.3.9104": denied: Unauthenticated request. Unauthenticated requests do not have permission "artifactregistry.repositories.downloadArtifacts" on resource "projects/sourcegraph-ci/locations/us-central1/repositories/rfc795-internal" (or it may not exist).
See 'docker run --help'.
--- 🕵️  Auto Upgrade Tests:
🚨 5.1.0 Failed -- 1m24.755906542s
🚨 candidate frontend error: frontend container timed out during polling version update:
Error response from daemon: No such container: auto_frontend_34794dc2
```

**Comment by jhchabran** [comment]
@DaedalusG yeah, for some reason, if we don't pull beforehand the images, they don't get pull automatically when running the containers.

File: testing/tools/upgradetest/tests-util.go [hunk 4]
```diff
@@ -221,6 +221,23 @@ func setupTestEnv(ctx context.Context, testType string, initVersion *semver.Vers
 	test.AddLog(fmt.Sprintf("Upgrading from version (%s) to release candidate.", initVersion))
 	test.AddLog("-- 🏗️  setting up test environment")
 
+	// Pull images from registry if -from-registry is set
+	if ctx.Value(fromRegistryKey{}).(string) != "sourcegraph/" {
```
Comment (DaedalusG) [inline testing/tools/upgradetest/tests-util.go:225]: I think here we really want to keep `targetRegistryKey` instead of `fromRegistryKey` it won't error because docker run is just pulling the images if they aren't available now, but if we want this I think the idea is that we keep it for the target images to be pulled early.

**Comment by jhchabran** [comment]
@DaedalusG it's all green over here: https://buildkite.com/sourcegraph/sourcegraph/builds/269364#018ee72c-422b-4049-81e1-728be53acf63

**Comment by DaedalusG** [comment]
@jhchabran added a little smarter exit logic and made the log output _really verbose_ for logging any failed test -- I forced a failure by overclocking things here: 
[testRun.txt](https://github.com/sourcegraph/sourcegraph/files/15020926/testRun.txt)

The passing case: 
[testRun2.txt](https://github.com/sourcegraph/sourcegraph/files/15021118/testRun2.txt)

This seems pretty good to go, but the logging on failure is pushing it on _too_ verbose maybe. wdyt?

**Review by DaedalusG — APPROVED** [review]

**Comment by jhchabran** [comment]
> This seems pretty good to go, but the logging on failure is pushing it on too verbose maybe. wdyt?

We can always revisit that later, as long as it fails the build, we're good to push forward, this stuff has been in flight for a while now.

**Review by jdpleiness — APPROVED** [review]
A few things come to mind, but @DaedalusG and I have another branch working on those improvements 🚀

File: release.yaml [hunk 5]
```diff
@@ -1,8 +1,8 @@
 meta:
   productName: sourcegraph
-  repository: "github.com/sourcegraph/sourcegraph"
+  repository: 'github.com/sourcegraph/sourcegraph'
```
Comment (BolajiOlajide) [inline release.yaml:3]: Curious, why the change from double to single quotes?

Comment (jhchabran) [reply release.yaml:3]: No reason that I can recall tbh

Comment (BolajiOlajide) [reply release.yaml:3]: Gotcha!
